core/permission/access.go

```diff
@@ -214,10 +214,12 @@ func (a Access) controllerValue() int {
 
 func (a Access) cloudValue() int {
 	switch a {
-	case AddModelAccess:
+	case NoAccess:
 		return 0
-	case AdminAccess:
+	case AddModelAccess:
 		return 1
+	case AdminAccess:
+		return 2
 	default:
 		return -1
 	}
```

Comment:
This seems dangerous... looking at the git history this has been like this for 6 years. Are we sure this is correct?

Reply:
I've added this as a separate commit with some reasoning:

> All other permissions can be compared against NoAccess. Cloud could not because the cloudValue function returned -1 for it and the EqualOrGreaterCloudAccessThan function returned false whenever things were -1.

All the other access levels include the NoAccess as the base. This `cloudValue` function is only used by `EqualOrGreaterCloudAccessThan`.

The current implementation would mean this returned:
```
(WriteAccess).EqualOrGreaterCloudAccessThan(NoAccess) == false
```
The function is not that widely used and `cloudValue` was added in an unreviewed PR.

As far as I could tell, in all the other use cases in the codebase apart from mine, the value passed as the second argument was fixed, and so it would never hit this case.

Fixing this makes the `EqualOrGreaterAccessThan` function behave consistently which is used in the code I'm adding.

Reply:
If that's the case we should pull this out of this PR and land it independently, with even the thought of landing it in 3.x.

Reply:
Put up a PR to land this in 3.4 https://github.com/juju/juju/pull/17800